org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/buildpath/BuildPathSupport.java

```diff
@@ -305,10 +305,10 @@ public boolean accept(File dir, String name) {
 			"org.junit", new VersionRange("[4.13.0,5.0.0)"), null, "org.junit_4.*.jar", "org.junit.source", "source-bundle/", JUnitPreferencesConstants.JUNIT4_JAVADOC); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$ //$NON-NLS-4$ //$NON-NLS-5$
 
 	private static final JUnitPluginDescription HAMCREST_CORE_PLUGIN= new JUnitPluginDescription(
-			"org.hamcrest.core", new VersionRange("[2.2.0,2.3.0)"), null, "org.hamcrest.core_2.*.jar", "org.hamcrest.core.source", "source-bundle/", JUnitPreferencesConstants.HAMCREST_CORE_JAVADOC); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$ //$NON-NLS-4$ //$NON-NLS-5$
+			"org.hamcrest.core", new VersionRange("[2.2.0,3.1.0)"), null, "org.hamcrest.core_2.*.jar", "org.hamcrest.core.source", "source-bundle/", JUnitPreferencesConstants.HAMCREST_CORE_JAVADOC); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$ //$NON-NLS-4$ //$NON-NLS-5$
```

Comment:
Will the same problem resurface when 3.1 is released?  It will happen in the future...

Reply:
I don't know why so strict ranges are used here. The usual construct is '[current.min, next.major)' and as 3.0 is good I would assume all 3.x versions should be good.

Reply:
Indeed my sense is we'd be better to assume that hamcrest will be well behaved with respect to semantic versions now that they are publishing OSGi bundles directly.

Reply:
It's sort of not empty because it contains this abomination re-exporting a bundle regardless of whether it might be version 1000.0.0 of that bundle:
```
Automatic-Module-Name                   org.hamcrest.core.deprecated
Bundle-ManifestVersion                  2
Bundle-Name                             Bundle org.hamcrest : hamcrest-core
Bundle-SymbolicName                     org.hamcrest.core
Bundle-Version                          2.2.0.v20230809-1000
Eclipse-Wrapped-Bundle                  org.hamcrest:hamcrest-core:2.2
Implementation-Title                    hamcrest-core
Implementation-Vendor                   hamcrest.org
Implementation-Version                  2.2
Import-Package                          java.lang
Manifest-Version                        1.0
Require-Bundle                          org.hamcrest;bundle-version="2.2.0";visibility:=reexport
Require-Capability                      osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.7))"
```

Reply:
> It's sort of not empty because it contains this abomination re-exporting a bundle regardless of whether it might be version 1000.0.0 of that bundle:

While this is relevant inside an OSGi framework it doesn't help for the classpath container because that one is just a collection of jars. So this entry actually is obsolete for 2.2... but we can remove it on the next cycle probably.

Reply:
> Does it actually fix the issue if 3.0 is NOT present?
> 
> As shown in [#1610 (comment)](https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/1610#issuecomment-2308793661) the 2.2.0 **seems to be completely empty** for JUNIT 5 so don't we need 3.x as _lower_ bound?


Yes, it does. The reason why the `JUnit-5` container does not fail completely is that for `JUnit-5`  library entries that are not found and therefore `null`, are removed, while for JUnit-4 they are preserved. So if a desired library is not found it leads to an NPE when initializing a `JUnit-4`container, while for `JUnit-5` the missing library is simply discarded, in this case hamcrest, which leads to compilation errors:
https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/b83ffe1bbcad393d4e35874d1ecc6af59f90874f/org.eclipse.jdt.junit.core/src/org/eclipse/jdt/internal/junit/buildpath/JUnitContainerInitializer.java#L150

> > It's sort of not empty because it contains this abomination re-exporting a bundle regardless of whether it might be version 1000.0.0 of that bundle:
> 
> While this is relevant inside an OSGi framework it doesn't help for the classpath container because that one is just a collection of jars. So this entry actually is obsolete for 2.2... but we can remove it on the next cycle probably.

That's right. Therefore I already created https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/1612.


> Will the same problem resurface when 3.1 is released? It will happen in the future...

That's right. But this PR is inteded to just fix the linked issue in the presence of hamcrest-3.0. I assume there won't be a 3.1 within the next three months.
And on the long run I think the fetching of libraries should be be made more dynamic any ways (I assume no one what's to bump the version all the time?).
I'm thinking about a 'declarative' way where one declares the 'anchor/root'-packages to include (maybe an 'anchor-class' if there are split-packages) and the Eclipse OSGi runtime is queried for the bundles providing it.

Reply:
We really need to move forward here for this release cycle so someone needs to push the magic button.  Things are definitely badly broken without a fix.  Please, please. 🙏